go/ql/src/experimental/CWE-321/HardcodedKeysLib.qll

```diff
@@ -0,0 +1,271 @@
+            "github.com/lestrrat-go/jwx", "github.com/lestrrat/go-jwx/jwk",
+            "github.com/lestrrat-go/jwx%/jwk"
+          ]) and
+        // `func New(key interface{}) (Key, error)`
+        m.getTarget().hasQualifiedName(pkg, "New")
+      |
+        this = m.getArgument(0)
+      )
+    }
+  }
+
+  /**
+   * Mark any comparision expression where any operand is tainted as a
+   * sanitizer for all instances of the taint
+   */
+  private class CompareExprSanitizer extends Sanitizer {
+    CompareExprSanitizer() {
+      exists(BinaryExpr c |
+        c.getAnOperand().getGlobalValueNumber() = this.asExpr().getGlobalValueNumber()
+      )
+    }
+  }
+
+  /** Mark an empty string returned with an error as a sanitizer */
+  class EmptyErrorSanitizer extends Sanitizer {
+    EmptyErrorSanitizer() {
+      exists(ReturnStmt r, DataFlow::CallNode c |
+        c.getTarget().hasQualifiedName("errors", "New") and
+        r.getNumChild() > 1 and
+        r.getAChild() = c.getAResult().getASuccessor*().asExpr() and
+        r.getAChild() = this.asExpr()
+      )
+    }
+  }
+
+  /** Mark any formatting string call as a sanitizer */
+  class FormattingSanitizer extends Sanitizer {
+    FormattingSanitizer() { exists(Formatting::StringFormatCall s | s.getAResult() = this) }
+  }
+
+  /**
+   * Mark any taint arising from a read on a tainted slice with a random index as a
+   * sanitizer for all instances of the taint
+   */
+  private class RandSliceSanitizer extends Sanitizer {
+    RandSliceSanitizer() {
+      exists(DataFlow::CallNode randint, string name, DataFlow::ElementReadNode r |
+        (
+          randint.getTarget().hasQualifiedName("math/rand", name) or
+          randint.getTarget().(Method).hasQualifiedName("math/rand", "Rand", name)
+        ) and
+        name =
+          [
+            "ExpFloat64", "Float32", "Float64", "Int", "Int31", "Int31n", "Int63", "Int63n", "Intn",
+            "NormFloat64", "Uint32", "Uint64"
+          ] and
+        r.reads(this, randint.getAResult().getASuccessor*())
+      )
+      or
+      // Sanitize flows like this:
+      // func GenerateCryptoString(n int) (string, error) {
+      //   const chars = "123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-"
+      //   ret := make([]byte, n)
+      //   for i := range ret {
+      //     num, err := crand.Int(crand.Reader, big.NewInt(int64(len(chars))))
+      //     if err != nil {
+      //       return "", err
+      //     }
+      //     ret[i] = chars[num.Int64()]
+      //   }
+      //   return string(ret), nil
+      // }
+      exists(
+        DataFlow::CallNode randint, DataFlow::MethodCallNode bigint, DataFlow::ElementReadNode r
+      |
+        randint.getTarget().hasQualifiedName("crypto/rand", "Int") and
+        bigint.getTarget().hasQualifiedName("math/big", "Int", "Int64") and
+        bigint.getReceiver() = randint.getResult(0).getASuccessor*() and
+        r.reads(this, bigint.getAResult().getASuccessor*())
```

Comment:
This is similar enough to the previous case, you could replace `getASuccessor*` with an augmented get-a-successor that includes the taint step across `big.Int.Int64` (or always include that as a taint step, and use localTaint like the case below)

Reply:
I am not sure I follow. Do you mean to say I should create a new taint tracking config and add this step there?

Reply:
No, I mean instead of using getASuccessor* aka localFlow to track flow from rand.Int() etc to Int.Int64() and then using it again to track from Int64() to an array-read, you could extend `AdditionalTaintStep` or `FunctionModel` to make `Int64()` taint-conserving and then use a single localTaint call to track flow from `rand.Int()` through to an array-read that you want to sanitize. Similarly you could make `%` modulus operations taint-conserving so they wouldn't have to be sanitized.

One other related question: it looks like in your sanitizers you sanitize the input to `constant[random]`, whereas logically you should probably sanitize the result instead.

The ideal version would look something like

using source = post-update-node of rand.Read(_) or result of rand.Int() etc, and additional taint steps = big.Int functions and modulus operations, sanitize track taint forwards to any index operand and sanitize that array-read's result.

Reply:
@smowton I think I partially understand what you wish to say. I have made some changes did you mean something on these lines?